runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -897,6 +900,104 @@ J9::TransformUtil::foldStaticFinalFieldAssumingProtection(TR::Compilation *comp,
    return false;
    }
 
+static bool
+classHasFinalPutstaticOutsideClinit(TR::Compilation *comp, TR_OpaqueClassBlock *clazz)
+   {
+   TR_J9VMBase *fej9 = comp->fej9();
+   J9VMThread *vmThread = fej9->vmThread();
+
+   J9ROMClass *romClass = TR::Compiler->cls.romClassOf(clazz);
+   TR::Region &stackRegion = comp->trMemory()->currentStackRegion();
+
+   // Get all of the final fields declared by clazz from the VM.
+   J9Class *j9c = TR::Compiler->cls.convertClassOffsetToClassPtr(clazz);
+   uintptr_t numStaticFields =
+      fej9->_vmFunctionTable->getStaticFields(vmThread, romClass, NULL);
+
+   uintptr_t staticFieldsSize = numStaticFields * sizeof(J9ROMFieldShape*);
+   J9ROMFieldShape **staticFields =
+      (J9ROMFieldShape**)stackRegion.allocate(staticFieldsSize);
+
+   fej9->_vmFunctionTable->getStaticFields(vmThread, romClass, staticFields);
+
+   // Partition so that the final ones are at the beginning and the non-final
+   // ones follow.
+   uintptr_t numStaticFinalFields = 0;
+   for (uintptr_t i = 0; i < numStaticFields; i++)
+      {
+      if ((staticFields[i]->modifiers & J9AccFinal) != 0)
+         {
+         // Field i is final. Swap to add it to the end of the final prefix.
+         // Swapping is OK because we don't depend on the order of the fields.
+         TR_ASSERT_FATAL(numStaticFinalFields < numStaticFields, "out of bounds");
+         J9ROMFieldShape *tmp = staticFields[numStaticFinalFields];
+         staticFields[numStaticFinalFields] = staticFields[i];
+         staticFields[i] = tmp;
+         numStaticFinalFields++;
+         }
+      }
+
+   if (numStaticFinalFields == 0)
+      return false;
+
+   // Iterate the bytecode of all methods looking for a putstatic to a static
```

Comment:
If the static final field is not private, is it enough to iterate the bytecodes of the methods in the declaring class ? i.e. is it possible that some other class writes to the static final field if it is public ?

Reply:
Yes, we only need to look at the declaring class regardless of visibility. If a `putstatic` instruction tries to store to a public static final field declared by another class, it will fail with IAE just as the spec says

The discrepancy was that `putstatic` instructions were allowed to store to static final fields declared by the current class from any of its methods even though the spec states that they can do so only directly from `<clinit>`

Reply:
[Here](https://github.com/eclipse-openj9/openj9/blob/9c173ce68e2e3fc4f10463dd5385e557011a211d/runtime/vm/resolvesupport.cpp#L768-L787) is where static field resolution checks to make sure that the current class is allowed to `putstatic` a static final field. It requires private access regardless of the declared visibility of the field. And with the no nestmates flag, private access is allowed only from the declaring class (or classes that are exempt from verification, but we're assuming that JCL classes are well-behaved)

Reply:
Thanks, I'll run testing on it while we wait for @mpirvu to review the JIT server part